### PR TITLE
[WFCORE-4750] Using regex for role in Elytron

### PR DESCRIPTION
**Pull request by Skyllarr**
Jira: [WFCORE-4750](https://issues.jboss.org/browse/WFCORE-4750)

Requires [ELY-1909](https://issues.jboss.org/browse/ELY-1909)

### CODE REVIEW

File: elytron/src/main/java/org/wildfly/extension/elytron/RoleMapperDefinitions.java [hunk 1]
```diff
@@ -80,6 +85,19 @@
         .setRestartAllServices()
         .build();
 
+    static final SimpleAttributeDefinition REGEX = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.REGEX, ModelType.STRING, false)
```
Comment (fjuma) [inline elytron/src/main/java/org/wildfly/extension/elytron/RoleMapperDefinitions.java]: I think you can use `SimpleAttributeDefinitionBuilder(RegexAttributeDefinitions.PATTERN)` here instead and then remove the `set` calls below.

File: elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
+package org.wildfly.extension.elytron;
```
Comment (fjuma) [inline elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java]: Missing copyright header.

File: elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        operation.get(ElytronDescriptionConstants.REPLACEMENT).set("$1");
+        ModelNode response = services.executeOperation(operation);
+        // operation will fail because regex is not valid (starts with asterisk)
+        if (! response.get(OUTCOME).asString().equals(FAILED)) {
+            Assert.fail(response.toJSONString(false));
+        }
+    }
+
+    @Before
+    public void init() throws Exception {
```
Comment (fjuma) [inline elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java]: Is this method actually needed?

File: elytron/src/main/java/org/wildfly/extension/elytron/ElytronDescriptionConstants.java [hunk 4]
```diff
@@ -427,7 +427,9 @@
     String REFERENCE = "reference";
     String REFERRAL_MODE = "referral-mode";
     String REGISTER_JASPI_FACTORY = "register-jaspi-factory";
+    String REGEX = "regex";
```
[the remark this answers is not in the document]
Comment (Skyllarr) [reply elytron/src/main/java/org/wildfly/extension/elytron/ElytronDescriptionConstants.java]: Fixed. Thanks

File: elytron/src/main/java/org/wildfly/extension/elytron/RoleMapperDefinitions.java (hunk 1 above)
Comment (Skyllarr) [reply elytron/src/main/java/org/wildfly/extension/elytron/RoleMapperDefinitions.java]: Fixed. Thank you

File: elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java (hunk 2 above)
Comment (Skyllarr) [reply elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java]: Fixed.

File: elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java (hunk 3 above)
Comment (Skyllarr) [reply elytron/src/test/java/org/wildfly/extension/elytron/RegexRoleMapperTestCase.java]: Fixed.

**Comment by fjuma** [comment]
Thanks, @Skyllarr. This looks good, it just has some conflicts that need to be resolved.

File: elytron/src/test/java/org/wildfly/extension/elytron/RoleMappersTestCase.java [hunk 5]
```diff
@@ -156,4 +167,216 @@ public void testKeepBothMappedRoleMapper() throws Exception {
...
+        context.setAuthenticationName("user5");
+        Assert.assertTrue(context.exists());
+        Assert.assertTrue(context.authorize());
+        context.succeed();
+        SecurityIdentity identity = context.getAuthorizedIdentity();
+        Assert.assertEquals("user5", identity.getPrincipal().getName());
+
+        Roles roles = identity.getRoles();
+        Assert.assertTrue(roles.contains("admin"));
+        Assert.assertTrue(roles.contains("admin"));
```
Comment (fabiobrz) [inline elytron/src/test/java/org/wildfly/extension/elytron/RoleMappersTestCase.java:365]: Maybe duplicated line here? From test domain definition I ca see that `user5` should have the following roles, resulting from `AggregateRoleMapper1` processing:
* `1-user-1`: match, translated into `admin`
* `joe`: no match, not kept
* `2-guest-2`: match, translated into `guest`

Hence looks like result should be `{admin, guest}`.
WDYT?

Comment (Skyllarr) [reply elytron/src/test/java/org/wildfly/extension/elytron/RoleMappersTestCase.java]: fixed

**Pull request merged**